### ISSUE TITLE
Code actions should support tab stops

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#356]
I would expect `widget` to be highlighted.

![Selection](https://user-images.githubusercontent.com/1078012/27956034-b8d59184-630f-11e7-979a-703e4b1b097b.png)

**Comment by DanTup** [comment]
Looks like we have the info required in `linkedEditGroups`.

```js
{
	"message": "Wrap with new widget",
	"edits": [
		{
			"file": "m:\\Coding\\Applications\\flutter\\examples\\stocks\\lib\\main.dart",
			"fileStamp": 0,
			"edits": [
				{
					"offset": 86,
					"length": 15,
					"replacement": "new widget(child: new Container())"
				}
			]
		}
	],
	"linkedEditGroups": [
		{
			"positions": [
				{
					"file": "m:\\Coding\\Applications\\flutter\\examples\\stocks\\lib\\main.dart",
					"offset": 90
				}
			],
			"length": 6,
			"suggestions": []
		},
		{
			"positions": [
				{
					"file": "m:\\Coding\\Applications\\flutter\\examples\\stocks\\lib\\main.dart",
					"offset": 97
				}
			],
			"length": 5,
			"suggestions": []
		}
	],
	"selection": {
		"file": "m:\\Coding\\Applications\\flutter\\examples\\stocks\\lib\\main.dart",
		"offset": 120
	}
}
```

**Comment by DanTup** [comment]
Can't see how to do this, have raised https://github.com/Microsoft/vscode/issues/30245

**Comment by DanTup** [comment]
https://code.visualstudio.com/docs/extensionAPI/vscode-api#TextEditor.insertSnippet

**Comment by DanTup** [comment]
Got a working POC on `code-action-tab-stops` branch; however there's some weird interaction with completion - when you hit `tab` to move between the placeholders it commits whatever is top of the completion list which is probably not what you want.

Test case:

```dart
library stocks;

import 'package:flutter/material.dart';

void main() => runApp(new Danny(thing: new Container()));


// test
// test
// test
// test
// test
// test
```

Invoke `Create class Danny` in the main method. When typing the class name then hitting tab, it'll just fill with something random.

**Comment by DanTup** [comment]
Possibly no current fixes use both multiple edits and `linkedEditGroups` but "Invert If statement" does use multiple edits. `Create function XYZ` populates suggestions.

**Comment by xster** [comment]
Not sure what the desired behaviour with tab is. 
Maybe this is splittable into steps. In the original screenshot in https://github.com/Dart-Code/Dart-Code/issues/356#issue-241235641, I'd be totally happy if doing wrap with new widget on `Container` just left the editor with `widget` selected. No follow up tabs or other operations supported. 

(since most typically, after replacing `widget`, users can just cmd+enter into the next line and add arguments between `widget` and its `child` argument)

**Comment by DanTup** [comment]
I don't remember all the details about this; I'll rebase the branch on latest code and do some testing to see where we are.

**Comment by DanTup** [comment]
I think the completion thing is a bug - for some reason on the second keypress Code invokes the completion provider. I've raised https://github.com/Microsoft/vscode/issues/43523

**Comment by DanTup** [comment]
I'm no longer convinced invoking the completion is a bad thing, since you might be completing something (eg. when you wrap in a widget, having completion on the widget name seems good).

Merging this into 2.9 as I think it works well.

**Comment by xster** [comment]
So what's the conclusion? Wrap a child widget and the selection is highlighting the parent `widget`? Further tab goes back to the child?

**Comment by DanTup** [comment]
When you invoke "Wrap in new widget" it'll do the usual thing, but highlight the text `widget` for you type over. If the completion is open while your'e typing there and you hit tab, it will commit the completion - this might actually be desired (eg. you want to complete the name of a widget class). The main reason I didn't like this previous was because of the parens being inserted, but I fixed that today in #564.

If the completion opens and you don't want to commit it, you'll need to hit escape or not use a key that commits it.

In my testing this all seemed fine; but ofc I'll revisit if it seems like we can do better. Hoping to push out a beta of v2.9 next week for anyone that wants to help test before it's released (I'll announce on twitter and in Gitter when it's available).
